lib/puppet/resource.rb

```diff
@@ -305,7 +305,7 @@ def resource_type
     when "Class"; environment.known_resource_types.hostclass(title == :main ? "" : title)
     when "Node"; environment.known_resource_types.node(title)
     else
-      Puppet::Type.type(type) || environment.known_resource_types.definition(type)
+      Puppet::Type.type(type) || environment.known_resource_types.definition(type) || environment.known_resource_types.application(type)
```

Comment:
This is performance sensitive (we create a lot of resources and every instruction counts). Suggest breaking out `environment.known_resource_types` to a `krt` variable instead of asking for it twice.